### PR TITLE
!str #19156 Source/Flow/Sink-Stage removed

### PR DESCRIPTION
**Pull request by johanandren**
Refs #19156

### CODE REVIEW

**Comment by johanandren** [comment]
Ping @rkuhn 

Does it need mentioning in the migration guide?

**Comment by rkuhn** [comment]
no, these are new

**Comment by ktoso** [comment]
If they were not documented I'd not explain in migration guide.

**Comment by hepin1989** [comment]
I think a better place maybe move them to `GraphStages` as helper class,I think we should need this sooner or later.

**Comment by viktorklang** [comment]
Given how infrequent writing custom GraphStages ought to be, I think having a canonical way to do so wins. i.e. :+1: from me in removing these.

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala [hunk 1]
```diff
@@ -93,8 +93,12 @@ private[http] object HttpServerBluePrint {
         .via(requestStartOrRunIgnore(settings)))
 
   def requestStartOrRunIgnore(settings: ServerSettings)(implicit mat: Materializer): Flow[(ParserOutput.RequestOutput, Source[ParserOutput.RequestOutput, Unit]), HttpRequest, Unit] =
-    Flow.fromGraph(new FlowStage[(RequestOutput, Source[RequestOutput, Unit]), HttpRequest, Unit]("RequestStartThenRunIgnore") {
-      override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = (new GraphStageLogic(shape) {
+    Flow.fromGraph(new GraphStage[FlowShape[(RequestOutput, Source[RequestOutput, Unit]), HttpRequest]] {
+      val in = Inlet[(RequestOutput, Source[RequestOutput, Unit])]("RequestStartThenRunIgnore.in")
+      val out = Outlet[HttpRequest]("RequestStartThenRunIgnore.out")
+      override def shape: FlowShape[(RequestOutput, Source[RequestOutput, Unit]), HttpRequest] = FlowShape.of(in, out)
```
Comment (viktorklang) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: Why is this a `def`? They were `val`s before.

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala [hunk 2]
```diff
@@ -33,7 +33,10 @@ private[akka] object InputStreamSinkStage {
 /**
  * INTERNAL API
  */
-private[akka] class InputStreamSinkStage(readTimeout: FiniteDuration) extends SinkStage[ByteString, InputStream]("InputStreamSink") {
+private[akka] class InputStreamSinkStage(readTimeout: FiniteDuration) extends GraphStageWithMaterializedValue[SinkShape[ByteString], InputStream] {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: `final class` please :)

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala [hunk 3]
```diff
@@ -33,7 +33,10 @@ private[akka] object InputStreamSinkStage {
...
+  val in = Inlet[ByteString]("InputStreamSink.in")
+  override def shape: SinkShape[ByteString] = SinkShape.of(in)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: val

File: akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala [hunk 4]
```diff
@@ -32,10 +32,13 @@ private[akka] object OutputStreamSourceStage {
   }
 }
 
-private[akka] class OutputStreamSourceStage(writeTimeout: FiniteDuration) extends SourceStage[ByteString, OutputStream]("OutputStreamSource") {
+private[akka] class OutputStreamSourceStage(writeTimeout: FiniteDuration) extends GraphStageWithMaterializedValue[SourceShape[ByteString], OutputStream] {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala]: `final class` please :)

File: akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala [hunk 5]
```diff
@@ -32,10 +32,13 @@ private[akka] object OutputStreamSourceStage {
...
   val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
   require(maxBuffer > 0, "Buffer size must be greater than 0")
 
+  val out = Outlet[ByteString]("OutputStreamSource.out")
+  override def shape: SourceShape[ByteString] = SourceShape.of(out)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala]: val

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala [hunk 6]
```diff
@@ -33,7 +33,10 @@ private[akka] object InputStreamSinkStage {
...
+
   val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
   require(maxBuffer > 0, "Buffer size must be greater than 0")
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: I propose to execute these two lines *at the very beginning* of the consutructor

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 7]
```diff
@@ -223,7 +228,12 @@ private[akka] final class LastOptionStage[T] extends SinkStage[T, Future[Option[
   }
 }
 
-private[akka] final class HeadOptionStage[T] extends SinkStage[T, Future[Option[T]]]("headOption") {
+private[akka] final class HeadOptionStage[T] extends GraphStageWithMaterializedValue[SinkShape[T], Future[Option[T]]] {
+
+  val in = Inlet[T]("headOption.in")
+
+  override def shape: SinkShape[T] = SinkShape.of(in)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: val

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 8]
```diff
@@ -193,7 +193,12 @@ private[akka] final class AcknowledgeSink[In](bufferSize: Int, val attributes: A
   override def toString: String = "AcknowledgeSink"
 }
 
-private[akka] final class LastOptionStage[T] extends SinkStage[T, Future[Option[T]]]("lastOption") {
+private[akka] final class LastOptionStage[T] extends GraphStageWithMaterializedValue[SinkShape[T], Future[Option[T]]] {
+
+  val in = Inlet[T]("lastOption.in")
+
+  override def shape: SinkShape[T] = SinkShape.of(in)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: val

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/StageActorRefSpec.scala [hunk 9]
```diff
@@ -178,7 +178,10 @@ object StageActorRefSpec {
 
   import ControlProtocol._
 
-  case class SumTestStage(probe: ActorRef) extends SinkStage[Int, Future[Int]]("IntSum") {
+  case class SumTestStage(probe: ActorRef) extends GraphStageWithMaterializedValue[SinkShape[Int], Future[Int]] {
+    val in = Inlet[Int]("IntSum.in")
+    override def shape: SinkShape[Int] = SinkShape.of(in)
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/StageActorRefSpec.scala]: val

**Comment by drewhk** [comment]
LGTM, these were not as useful

**Comment by rkuhn** [comment]
LGTM (but still has compilation errors)

**Comment by drewhk** [comment]
yep, inheritance from final classes

**Comment by johanandren** [comment]
The @viktorklang made me do it!

**Comment by viktorklang** [comment]
I did not make you push without having compiled.

-- 
Cheers,
√

**Comment by johanandren** [comment]
Argh, the `val shape` is not a good idea as it is used by the GraphStage constructor. What is better in this case, `lazy val` or `def`?

**Comment by johanandren** [comment]
Oh, wait, it was the ordering of the lines that made it, module depends on shape.

**Comment by ktoso** [comment]
LGTM once passes; please squash :+1:

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala (hunk 2 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: ping here (final please)

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: It has a subclass in test

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: oh ok, that's the final you mean there.

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: Pardon my ignorance, but is that subclass *strictly* needed?

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: Consider it low hanging fruit that I'm friendly enough that I leave for someone else to fix. ;)

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: You can ticketify then.

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: #19170

**Pull request merged**